tests/functional/models/test_application.py

```diff
@@ -0,0 +1,145 @@
+
+    def test_enable_rolling_updates(self):
+        # should enable rolling update for the applications devices.
+        app = json.loads(self.resin.models.application.create('FooBar', 'Raspberry Pi 2').decode('utf-8'))
+        self.resin.models.application.enable_rolling_updates(app['id'])
+        app = self.resin.models.application.get('FooBar')
+        self.assertTrue(app['should_track_latest_release'])
+
+    def test_disable_rolling_updates(self):
+        # should disable rolling update for the applications devices.
+        app = json.loads(self.resin.models.application.create('FooBar', 'Raspberry Pi 2').decode('utf-8'))
+        self.resin.models.application.disable_rolling_updates(app['id'])
+        app = self.resin.models.application.get('FooBar')
+        self.assertFalse(app['should_track_latest_release'])
+
+    def test_enable_device_urls(self):
+        # should enable the device url for the applications devices.
+        app = json.loads(self.resin.models.application.create('FooBar', 'Raspberry Pi 2').decode('utf-8'))
+        device = json.loads(self.resin.models.device.register(app['id'], self.resin.models.device.generate_uuid()).decode('utf-8'))
+        self.resin.models.application.enable_device_urls(app['id'])
+        self.assertTrue(self.resin.models.device.has_device_url(device['uuid']))
+
+    def test_disable_device_urls(self):
+        # should disable the device url for the applications devices.
+        app = json.loads(self.resin.models.application.create('FooBar', 'Raspberry Pi 2').decode('utf-8'))
+        device = json.loads(self.resin.models.device.register(app['id'], self.resin.models.device.generate_uuid()).decode('utf-8'))
+        self.resin.models.application.enable_device_urls(app['id'])
+        self.resin.models.application.disable_device_urls(app['id'])
+        self.assertFalse(self.resin.models.device.has_device_url(device['uuid']))
+
+    def test_grant_support_access(self):
+        app = json.loads(self.resin.models.application.create('FooBar', 'Raspberry Pi 2').decode('utf-8'))
+        # should throw an error if the expiry timestamp is in the past.
+        expiry_timestamp = int((datetime.utcnow() - datetime.utcfromtimestamp(0)).total_seconds() * 1000 - 10000)
+        with self.assertRaises(self.helper.resin_exceptions.InvalidParameter):
+            self.resin.models.application.grant_support_access(app['id'], expiry_timestamp)
+
+        # should grant support access until the specified time.
+        expiry_time = int((datetime.utcnow() - datetime.utcfromtimestamp(0)).total_seconds() * 1000 + 3600 * 1000)
+        self.resin.models.application.grant_support_access(app['id'], expiry_time)
+        support_date = datetime.strptime(self.resin.models.application.get('FooBar')['is_accessible_by_support_until__date'], '%Y-%m-%dT%H:%M:%S.%fZ')
+        self.assertEqual(int((support_date - datetime.utcfromtimestamp(0)).total_seconds() * 1000), expiry_time)
```

Comment:
Might make sense to tweak this to:
```python
        support_expiry_time = int((support_date - datetime.utcfromtimestamp(0)).total_seconds() * 1000)
        self.assertEqual(support_expiry_time, expiry_time)
```
(IMHO it makes it slightly more obvious what's going on, and this'll probably be easier to debug should any problems arise)

Reply:
Or in fact, perhaps adding
```python
def datetime_to_epoch_ms(datevalue):
    return int((datevalue - datetime.utcfromtimestamp(0)).total_seconds() * 1000)
```
would be useful, as there are several places in these unit-tests where that could be used :-)

Reply:
Haha, which is also what https://stackoverflow.com/a/11111177 says ;)

Reply:
👍 Updated